merkle/merkle_queue.go

```diff
@@ -0,0 +1,325 @@
+		right = createQueueElement(right.depth-1, hash, rlpHash)
+		merkle.proofState.store(right)
+	}
+}
+
+func (merkle *merkleQueue) getPrecomputedHash(elementDepth int32) util.Hash {
+	diff := maxDepth - 1 - merkle.depth
+	return precomputedHashes[elementDepth + diff]
+}
+
+func (merkle *merkleQueue) computeRootHashes() error {
+	merkle.cancel()
+	merkle.Lock()
+	defer merkle.Unlock()
+
+	if merkle.leaves == 0 {
+		merkle.current[0] = createQueueElement(merkle.depth,
+			merkle.getPrecomputedHash(0),
+			merkle.getPrecomputedHash(0))
+		return nil
+	}
+
+	var depth = merkle.depth
+	for ; depth >= 0; depth-- {
+		if merkle.current[depth] != nil {
+			break
+		}
+	}
+	if depth <= 0 {
+		return nil
+	}
+
+	right := createQueueElement(depth,
+		merkle.getPrecomputedHash(depth),
+		merkle.getPrecomputedHash(depth))
+	merkle.proofState.store(right)
+
+	var left *queueElement
+	for ;depth > 0; depth-- {
+		left = merkle.current[depth] // this should non-null for the first iteration
+		/*
+		If there's no sibling, use a precomputed one.
+		Also swap left and right, as precomputed values are on the right-hand side.
+		 */
+		if left == nil {
+			left  = right
+			right = createQueueElement(depth,
+				merkle.getPrecomputedHash(depth),
+				merkle.getPrecomputedHash(depth))
+			merkle.proofState.store(right)
+		}
+		merkle.current[depth] = nil
+
+		right = createQueueElement(depth-1,
+			merkle.hasher(append(left.hash, right.hash...)),
+			merkle.hasher(append(left.rlpHash, right.rlpHash...)))
+		merkle.proofState.store(right)
+	}
+	merkle.current[0] = right
+
+	return nil
+}
+
+func (merkle *merkleQueue) listen() {
+	merkle.Lock()
+	defer merkle.Unlock()
+	for {
+		select {
+		case tx := <-merkle.queue:
+			tx.SetIndex(merkle.leafIndex)
+			merkle.leafIndex++
+			newElement := createQueueElement(merkle.depth, tx.Hash(), tx.RLPHash())
+			merkle.proofState.store(newElement)
+			merkle.processElement(newElement)
+
+		case <-merkle.context.Done():
+			return
+		}
+	}
+}
+
+func (merkle *merkleQueue) Reset() {
+	for i := int32(0); i < merkle.depth; i++ {
+		merkle.current[i] = nil
+	}
+	merkle.leaves = uint32(0)
+	merkle.leafIndex = uint32(0)
+	merkle.context, merkle.cancel = context.WithCancel(context.Background())
+
+	if merkle.proofIndex >= 0 {
+		merkle.proofState = createProofState(merkle.proofIndex, merkle.depth)
+	}
+	go merkle.listen()
+}
+
+func (merkle *merkleQueue) Enqueue(transactions []DualHashable) error {
+	if float64(merkle.leaves + uint32(len(transactions))) > merkle.maxLeaves {
+		return errors.New("Max number of leaves reached")
+	}
+	for i := 0; i < len(transactions); i++ {
+		merkle.leaves++
+		merkle.queue <- transactions[i]
+	}
+
+	return nil
+}
+
+func (merkle *merkleQueue) GetRootHash() (util.Hash, error) {
+	err := merkle.computeRootHashes()
+	if err != nil {
+		return nil, err
+	}
+	return merkle.current[0].hash, nil
+}
+
+func (merkle *merkleQueue) GetRootRLPHash() (util.Hash, error) {
+	err := merkle.computeRootHashes()
```

Comment:
can these `Get` methods be called multiple times?

Reply:
Yes, it can be called multiple times. However, after the first call the channel is stopped and adding more elements to the tree won't work.